src/Symfony/Component/Process/Process.php

```diff
@@ -1562,4 +1565,33 @@ private function escapeArgument($argument)
 
         return '"'.str_replace(array('"', '^', '%', '!', "\n"), array('""', '"^^"', '"^%"', '"^!"', '!LF!'), $argument).'"';
     }
+
+    private function replacePlaceholder($commandLine, $windows, $env)
+    {
+        $toReplace = $windows ? array('!', '!') : array('$', '');
+        if (!preg_match_all('/\{([a-zA-Z0-9]+?)\}/', $commandLine, $matches)) {
+            return $commandLine;
+        }
+
+        $patterns = $replacements = array();
+        $isInEnv = false;
+        foreach ($matches as $match) {
+            $patterns[] = sprintf('/{%s}/', $match[0]);
+            $replacements[] = sprintf('%s%s%s', $toReplace[0], $match[0], $toReplace[1]);
+            if (isset($env[$match[0]])) {
+                $isInEnv = true;
+                unset($env[$match[0]]);
+            }
+        }
+
+        if (!$isInEnv) {
+            throw new InvalidArgumentException(sprintf('There are no passed env for the command line "%s", replacement cannot be done.', $commandLine));
+        }
+
+        if (!empty($env)) {
+            throw new InvalidArgumentException(sprintf('Placeholder%s "%s" doesn\'t exist.', count($env) > 1 ? 's' : '', implode(',', array_keys($env))));
```

Comment:
If you make the distinction between singular and plural for "placeholder/s", you should probably do the same for "doesn't/don't" too.

Reply:
done